File: android/app/src/main/java/com/cuscatlan/pokemon/presentation/viewmodels/PokemonListViewModel.kt
package com.cuscatlan.pokemon.presentation.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.cuscatlan.pokemon.data.model.PokemonDisplayData
import com.cuscatlan.pokemon.data.model.PokemonListItem
import com.cuscatlan.pokemon.data.repository.PokemonRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class PokemonListViewModel(
    private val repository: PokemonRepository
) : ViewModel() {
    
    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()
    
    private val _pokemonList = MutableStateFlow<List<PokemonListItem>>(emptyList())
    val pokemonList: StateFlow<List<PokemonListItem>> = _pokemonList.asStateFlow()
    
    private val _pokemonDetails = MutableStateFlow<Map<Int, PokemonDisplayData>>(emptyMap())
    val pokemonDetails: StateFlow<Map<Int, PokemonDisplayData>> = _pokemonDetails.asStateFlow()
    
    private val _selectedPokemon = MutableStateFlow<Set<Int>>(emptySet())
    val selectedPokemon: StateFlow<Set<Int>> = _selectedPokemon.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    
    private val _loadingPokemonIds = MutableStateFlow<Set<Int>>(emptySet())
    val loadingPokemonIds: StateFlow<Set<Int>> = _loadingPokemonIds.asStateFlow()
    
    val filteredPokemonList: StateFlow<List<PokemonListItem>> = combine(
        _pokemonList,
        _searchQuery
    ) { list, query ->
        if (query.isBlank()) {
            list
        } else {
            list.filter { pokemon ->
                pokemon.name.contains(query, ignoreCase = true) || 
                pokemon.id.toString().contains(query)
            }
        }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = emptyList()
    )
    
    val selectedPokemonData: StateFlow<List<PokemonDisplayData>> = combine(
        _selectedPokemon,
        _pokemonDetails
    ) { selected, details ->
        selected.mapNotNull { id -> details[id] }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = emptyList()
    )
    
    init {
        loadPokemonList()
    }
    
    private fun loadPokemonList() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            
            repository.getPokemonList()
                .onSuccess { pokemonList ->
                    _pokemonList.value = pokemonList
                }
                .onFailure { throwable ->
                    _error.value = "Failed to load Pokémon list: ${throwable.message}"
                }
            
            _isLoading.value = false
        }
    }
    
    fun updateSearchQuery(query: String) {
        _searchQuery.value = query
    }
    
    fun loadPokemonDetails(pokemonId: Int) {
        if (_pokemonDetails.value.containsKey(pokemonId) || 
            _loadingPokemonIds.value.contains(pokemonId)) {
            return
        }
        
        viewModelScope.launch {
            _loadingPokemonIds.value = _loadingPokemonIds.value + pokemonId
            
            repository.getPokemonDetails(pokemonId)
                .onSuccess { pokemonData ->
                    _pokemonDetails.value = _pokemonDetails.value + (pokemonId to pokemonData)
                }
                .onFailure { throwable ->
                    _error.value = "Failed to load Pokémon details: ${throwable.message}"
                }
            
            _loadingPokemonIds.value = _loadingPokemonIds.value - pokemonId
        }
    }
    
    fun togglePokemonSelection(pokemonId: Int) {
        val currentSelection = _selectedPokemon.value
        val newSelection = if (currentSelection.contains(pokemonId)) {
            currentSelection - pokemonId
        } else {
            if (currentSelection.size < 3) {
                currentSelection + pokemonId
            } else {
                currentSelection // Don't add if already at limit
            }
        }
        _selectedPokemon.value = newSelection
    }
    
    fun clearSelection() {
        _selectedPokemon.value = emptySet()
    }
    
    fun clearError() {
        _error.value = null
    }
    
    fun retry() {
        loadPokemonList()
    }
    
    fun setSelectedPokemon(pokemonIds: Set<Int>) {
        _selectedPokemon.value = pokemonIds
    }
}

File: ios/pokedex-cuscatlan/ViewModels/PokemonListViewModel.swift
//
//  PokemonListViewModel.swift
//  pokedex-cuscatlan
//
//  Created by User on 16/9/25.
//

import Foundation
import Combine

class PokemonListViewModel: ObservableObject {
    @Published var pokemonList: [PokemonListItem] = []
    @Published var pokemonDetails: [Int: PokemonDisplayData] = [:]
    @Published var filteredPokemonList: [PokemonListItem] = []
    @Published var searchQuery: String = "" {
        didSet {
            filterPokemon()
        }
    }
    @Published var isLoading = false
    @Published var error: String?
    @Published var selectedPokemon: Set<Int> = []
    @Published var selectedPokemonData: [PokemonDisplayData] = []

    private let repository = PokemonRepository.shared
    private var cancellables = Set<AnyCancellable>()

    init() {
        loadPokemonList()
    }

    func loadPokemonList() {
        isLoading = true
        error = nil
        print("🔄 Loading Pokemon list...")

        repository.fetchPokemonList()
            .sink(
                receiveCompletion: { [weak self] completion in
                    DispatchQueue.main.async {
                        self?.isLoading = false
                        if case .failure(let error) = completion {
                            print("❌ Error loading Pokemon list: \(error.localizedDescription)")
                            self?.error = "Error loading Pokémon: \(error.localizedDescription)"
                        } else {
                            print("✅ Pokemon list loaded successfully")
                        }
                    }
                },
                receiveValue: { [weak self] pokemonList in
                    DispatchQueue.main.async {
                        print("📝 Received \(pokemonList.count) Pokemon")
                        self?.pokemonList = pokemonList
                        self?.filteredPokemonList = pokemonList
                    }
                }
            )
            .store(in: &cancellables)
    }

    func loadPokemonDetails(id: Int) {
        guard pokemonDetails[id] == nil else { return }

        repository.fetchPokemonDetail(id: id)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        DispatchQueue.main.async {
                            self?.error = "Error loading Pokémon details: \(error.localizedDescription)"
                        }
                    }
                },
                receiveValue: { [weak self] pokemonData in
                    DispatchQueue.main.async {
                        self?.pokemonDetails[id] = pokemonData
                    }
                }
            )
            .store(in: &cancellables)
    }

    func togglePokemonSelection(_ id: Int) {
        if selectedPokemon.contains(id) {
            selectedPokemon.remove(id)
        } else if selectedPokemon.count < 3 {
            selectedPokemon.insert(id)
        }
        updateSelectedPokemonData()
    }

    func clearSelection() {
        selectedPokemon.removeAll()
        updateSelectedPokemonData()
    }

    private func updateSelectedPokemonData() {
        selectedPokemonData = selectedPokemon.compactMap { pokemonDetails[$0] }
    }

    private func filterPokemon() {
        if searchQuery.isEmpty {
            filteredPokemonList = pokemonList
        } else {
            filteredPokemonList = pokemonList.filter { pokemon in
                pokemon.name.lowercased().contains(searchQuery.lowercased()) ||
                String(pokemon.id).contains(searchQuery)
            }
        }
    }

    func updateSearchQuery(_ query: String) {
        searchQuery = query
    }

    func clearError() {
        error = nil
    }

    func retry() {
        loadPokemonList()
    }
}